tests/test_data_utils/test_downsample_cells.py

```diff
@@ -1,98 +1,150 @@
+    def check_stratified_counts(
+            self, downsampled_df, columns, expected_counts):
+        """
+        Helper method to check stratified counts for both single and
+        multi annotations.
+        """
+        combined_col = '_'.join(columns)
 
-    def test_downsample_no_n_samples(self):
-        """Downsample without n_samples should return the original dataframe"""
-        df_downsampled = downsample_cells(self.df, self.annotation)
-        pd.testing.assert_frame_equal(self.df, df_downsampled)
+        # Ensure the combined column exists in the downsampled_df
+        if combined_col not in downsampled_df.columns:
+            downsampled_df[combined_col] = downsampled_df[columns].apply(
+                lambda row: '_'.join(row.values.astype(str)), axis=1
+            )
 
-    def test_stratification_frequency(self):
-        """Test the frequency of stratification"""
-        n_samples = 10
+        actual_counts = downsampled_df[combined_col].value_counts()
+
+        for key, expected_count in expected_counts.items():
+            self.assertEqual(expected_count, actual_counts.get(key, 0))
+
+        # Remove the combined column after checks
+        if combined_col in downsampled_df.columns and len(columns) > 1:
+            # Only drop if it's a multi annotation
+            downsampled_df.drop(columns=combined_col, inplace=True)
 
-        # Downsampling with stratifying
+    def test_stratification_frequency_single_obs(self):
+        """Test the frequency of stratification for single annotation."""
+        n_samples = 10
         df_downsampled_stratified = downsample_cells(
-            self.df, self.annotation,
+            self.df_single_fixed, self.annotation_single,
             n_samples=n_samples, stratify=True, rand=True
         )
-
-        # Assert the total number of samples
-        num_rows = df_downsampled_stratified.shape[0]
-        self.assertEqual(
-            num_rows, n_samples,
-            'Number of rows in df_downsampled_stratified does not match '
-            'n_samples'
+        expected_counts = {
+            'annotation1': 6,
+            'annotation2': 3,
+            'annotation3': 1
+        }
+        self.check_stratified_counts(
+            df_downsampled_stratified,
+            [self.annotation_single],
+            expected_counts
         )
 
-        # Assert the number of samples in each annotation
-        annotation = 'annotation'
-        annotation_value_counts = \
-            df_downsampled_stratified[annotation].value_counts()
-        self.assertEqual(
-            annotation_value_counts['annotation1'], 5,
-            'Number of samples in annotation1 does not match expected count'
-        )
-        self.assertEqual(
-            annotation_value_counts['annotation2'], 3,
-            'Number of samples in annotation2 does not match expected count'
+    def test_stratification_frequency_multi_obs(self):
```

Comment:
@fangliu117 , would you like to add a test case where there is an actual downsampling (e.g., start with with larger number of cells).

Reply:
yes, a large dataset for testing multiple annotation was applied.